### DISCUSSION TITLE
Adafruit ILI9341 TFT issue

### DISCUSSION THREAD

**Discussion by kronks6** [earlephilhower/arduino-pico#1612] [General]
Hello,

I am using this core for a project using spi0 on an Adafruit ILI9341 device with the following pins:
MOSI : GP19
MISO : GP16 
SCK : GP18
CS : GP17
RST,DC,BL : GP20-22 

Using Adafruits library, the default speed is very slow - drawing my starting page takes more than 1 second. It uses the drawRGBBitmap function to build a pretty GUI and these elements seem to take the longest, but everything else is much slower than I think it should be at what I thought would be 64MHz. The rest of the program includes a few buttons, PWM to drive motors, one interrupt pin for a flow meter, ADC, and some LED GPIO. The first try, I initialized the TFT with the whole pinout. I tried to run tft.begin(64000000) with no luck.  I tried using the setXX() function for the pins. When I tried to initialize the TFT with the SPIClass - the pico "crashed" and the program suspended. I had a LED turn on after initializing the screen and Serial.begin() was never called (no COM port showed up), so I know the program stopped running in some sense. I modified the program to work with Bodmer's TFT_eSPI, the exact same issue occurred and the program stopped running. 

I am going to try to modify the program further and take out features to find the cause, as well as benchmark SPI in the program and a separate test program. I have not gone super far into it yet, started porting the code to a Pico board this morning, but this seems like it could be a deeper issue with the core to me.

**Comment by maxgerhardt** [comment]
Did you resolve this? Just for my interest I read through 

https://github.com/adafruit/Adafruit_ILI9341/blob/93493769e9f55e26356fae93b94c9235e03888a8/Adafruit_ILI9341.h#L134-L144

and

https://github.com/adafruit/Adafruit-GFX-Library/blob/83a182d26011063411d59ec97d65c41d2d6ebb59/Adafruit_SPITFT.h#L125-L147

And it seems if you don't construct the obejct as `Adafruit_SPITFT (widht, height, &SPI, chipselect, dc, rst);`, then you get a very slow software SPI. You'll wanna use the hardware SPI for better speed.

**Reply by kronks6** [reply]
Hi Max, 

Thank you for the reply! I recognize your name from working with this kind of stuff, thank you for all you do.

I resolved the slow SPI, my issue was calling tft.begin() in a class constructor. I was able to do this on a different micro (Arduino Portenta) I used for development, but perhaps it was slower than it could have been. I never actually tested my fix with Adafruit's lib to see if that worked - I switched to TFT_eSPI while trying to debug and stuck with it.
